<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8">

<title>Bivalency</title>
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/4.7.0/css/font-awesome.min.css">
<link rel="stylesheet" href="./index_files/style.css">
</head>

<body>

<div class="container">
	<div class="content">
		<h1>Rethinking Bivalency in Early Differentiation</h1>
		<h4>EpiCypher 2021: Biological and Clinical Frontiers in Epigenetics</h4>
		<hr>
		<div class="bannerholder">
		  <img src="./index_files/home_banner.jpg">
		</div>
		<hr>
		<h2 id = "intro"><h2>Preprint and Data</h2>
		<p>If you would like to read our preprint, it is available at <a href="https://www.biorxiv.org/content/10.1101/2021.09.09.458948v1">bioRxiv</a>. The ICeChIP-seq data generated for this study is available at the Gene Expression Omnibus <a href="https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE108747">here</a> and <a href="https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE183155">here</a>.</p>

		<hr>
		
		<h2 id = "abstract">Abstract</h2>
		<p>Nucleosomes, composed of DNA and histone proteins, represent the fundamental repeating unit of the eukaryotic genome; post-translational modifications of these histone proteins influence the activity of the associated genomic regions to regulate cell identity. Traditionally, trimethylation of histone 3-lysine 4 (H3K4me3) is associated with transcriptional initiation, whereas trimethylation of H3K27 (H3K27me3) is considered transcriptionally repressive. The apparent juxtaposition of these opposing marks, termed “bivalent domains,” was proposed to specifically demarcate a small set of transcriptionally-poised lineage-commitment genes that resolve to one constituent modification through differentiation, thereby determining transcriptional status. Since then, many thousands of studies have canonized the bivalency model as a chromatin hallmark of development in many cell types. However, these conclusions are largely based on chromatin immunoprecipitations (ChIP) with significant methodological problems hampering their interpretation. Absent direct quantitative measurements, it has been difficult to evaluate the strength of the bivalency model. Here, we present reICeChIP, a calibrated sequential ChIP method to quantitatively measure H3K4me3/H3K27me3 bivalency genome-wide, addressing the limitations of prior measurements. With reICeChIP, we profile bivalency through the differentiation paradigm that first established this model: from naïve mouse embryonic stem cells (mESCs) into neuronal progenitor cells (NPCs). Our results cast doubt on every aspect of the bivalency model; in this context, we find that bivalency is widespread, does not resolve with differentiation, and is neither sensitive nor specific for identifying poised developmental genes or gene expression status more broadly. Our findings caution against interpreting bivalent domains as specific markers of developmentally poised genes.</p>
				
		<hr>
		
		<h2 id="contact">Contact and Acknowlegements</h2>
		<p>If you have any questions, comments, or issues, please feel free to contact Rohan Shah (rohanshah@uchicago.edu) or Alex Ruthenburg (aruthenburg@uchicago.edu).</p>
		<p>If you would like to see my other projects or download my CV, you can visit my website at <a href="https://shah-rohan.github.io/">https://shah-rohan.github.io/</a>.</p>
		<p>You can also visit the Ruthenburg Lab website to see some of the other work we are doing at <a href="https://ruthenlab.org/">https://ruthenlab.org/</a>.</p>
	</div>
</div>

</body></html>
